pkg/metrics/metrics.go

```diff
@@ -1,143 +1,328 @@
-		PolicyRuleInfo:          policyRuleInfoMetric,
-		PolicyChanges:           policyChangesMetric,
-		PolicyExecutionDuration: policyExecutionDurationMetric,
-		AdmissionReviewDuration: admissionReviewDurationMetric,
-		AdmissionRequests:       admissionRequestsMetric,
-	}
-
-	pc.MetricsRegistry.MustRegister(pc.Metrics.PolicyResults)
-	pc.MetricsRegistry.MustRegister(pc.Metrics.PolicyRuleInfo)
-	pc.MetricsRegistry.MustRegister(pc.Metrics.PolicyChanges)
-	pc.MetricsRegistry.MustRegister(pc.Metrics.PolicyExecutionDuration)
-	pc.MetricsRegistry.MustRegister(pc.Metrics.AdmissionReviewDuration)
-	pc.MetricsRegistry.MustRegister(pc.Metrics.AdmissionRequests)
-
-	// configuring metrics periodic refresh
-	if pc.Config.GetMetricsRefreshInterval() != 0 {
-		if len(pc.cron.Entries()) > 0 {
-			logger.Info("Skipping the configuration of metrics refresh. Already found cron expiration to be set.")
-		} else {
-			_, err := pc.cron.AddFunc(fmt.Sprintf("@every %s", pc.Config.GetMetricsRefreshInterval()), func() {
-				logger.Info("Resetting the metrics as per their periodic refresh")
-				pc.Metrics.PolicyResults.Reset()
-				pc.Metrics.PolicyRuleInfo.Reset()
-				pc.Metrics.PolicyChanges.Reset()
-				pc.Metrics.PolicyExecutionDuration.Reset()
-				pc.Metrics.AdmissionReviewDuration.Reset()
-				pc.Metrics.AdmissionRequests.Reset()
-			})
-			if err != nil {
-				return nil, err
-			}
-			logger.Info(fmt.Sprintf("Configuring metrics refresh at a periodic rate of %s", pc.Config.GetMetricsRefreshInterval()))
-			pc.cron.Start()
-		}
-	} else {
-		logger.Info("Skipping the configuration of metrics refresh as 'metricsRefreshInterval' wasn't specified in values.yaml at the time of installing kyverno")
+	c := controller.New(
+		processor.NewFactory(
+			simple.NewWithHistogramDistribution(),
+			aggregation.CumulativeTemporalitySelector(),
+			processor.WithMemory(true),
+		),
+		controller.WithResource(res),
+	)
+
+	exporter, err := prometheus.New(config, c)
+	if err != nil {
+		log.Error(err, "failed to initialize prometheus exporter")
+		return nil, nil, err
 	}
-	return pc, nil
+
+	global.SetMeterProvider(exporter.MeterProvider())
+
+	// Create new config object and attach metricsConfig to it
+	m := new(MetricsConfig)
+	m.Config = metricsConfigData
+
+	// Initialize metrics logger
+	m.Log = log
+
+	m, err = initializeMetrics(m)
+
+	if err != nil {
+		log.Error(err, "failed to initialize metrics config")
+	}
+
+	metricsServerMux := http.NewServeMux()
+	metricsServerMux.HandleFunc("/metrics", exporter.ServeHTTP)
+
+	return m, metricsServerMux, nil
+}
+
+func (m *MetricsConfig) RecordPolicyResults(policyValidationMode PolicyValidationMode, policyType PolicyType, policyBackgroundMode PolicyBackgroundMode, policyNamespace string, policyName string,
+	resourceKind string, resourceNamespace string, resourceRequestOperation ResourceRequestOperation, ruleName string, ruleResult RuleResult, ruleType RuleType,
+	ruleExecutionCause RuleExecutionCause, log logr.Logger) {
```

Comment:
`MetricsConfig` has a logger, do we need to pass it in?

Reply:
Removed this.